src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs

```diff
@@ -171,22 +177,31 @@ private TypeSymbol GetBestType(ImmutableArray<TypeSymbol> types, ref HashSet<Dia
             // that every type *before* best was also worse.
             for (int i = 0; i < bestIndex; i++)
             {
-                TypeSymbol type = types[i];
-                TypeSymbol better = Better(best, type, ref useSiteDiagnostics);
+                var type = types[i];
+                var better = Better(best, type, ref useSiteDiagnostics);
 
-                if (better != best)
+                if (!best.Equals(better, TypeCompareKind.ConsiderEverything))
                 {
                     return null;
                 }
             }
 
+            // If any of the types are null, the result should be nullable.
```

Comment:
I'm concerned about using `t is null` to identify `null` expressions. What about lambdas or typeless tuples `(null, 1)`?